### ISSUE TITLE
Infinite recursion due to `lib/modules.nix` changes

### ISSUE THREAD

**Issue by KaiSforza** [NixOS/nixpkgs#7250]
I have been attempting to build a system using the git repository, and was successful earlier, but had not pulled down the latest changes. After pulling down changes today I'm getting an error with infinite recursion. It seems to be coming from the `lib/modules.nix`. Below is the error that I was seeing.

Running a `git bisect` between the version that worked for me 14ea4e7e7 and the current `master`, 1035b7b6c9, I ended up at 1d62ad474 as the probable culprit.

```
# nixos-rebuild build --show-trace -I nixpkgs=/root/nixpkgs
building Nix...
error: while evaluating the attribute ‘config’ at "/root/nixpkgs/nixos/lib/eval-config.nix":26:5:
while evaluating anonymous function at "/root/nixpkgs/pkgs/top-level/all-packages.nix":8:1, called from "/root/nixpkgs/nixos/lib/eval-config.nix":55:10:
while evaluating ‘applyGlobalOverrides’ at "/root/nixpkgs/pkgs/top-level/all-packages.nix":107:26, called from "/root/nixpkgs/pkgs/top-level/all-packages.nix":101:10:
while evaluating ‘pkgsFun’ at "/root/nixpkgs/pkgs/top-level/all-packages.nix":125:19, called from "/root/nixpkgs/pkgs/top-level/all-packages.nix":120:14:
while evaluating the attribute ‘config.nixpkgs’ at "/root/nixpkgs/nixos/lib/eval-config.nix":26:5:
while evaluating the attribute ‘config’ at "/root/nixpkgs/lib/modules.nix":56:25:
while evaluating ‘yieldConfig’ at "/root/nixpkgs/lib/modules.nix":43:29, called from "/root/nixpkgs/lib/modules.nix":42:16:
while evaluating ‘mapAttrs’ at "/root/nixpkgs/lib/attrsets.nix":134:17, called from "/root/nixpkgs/lib/modules.nix":44:32:
while evaluating ‘mergeModules’ at "/root/nixpkgs/lib/modules.nix":111:26, called from "/root/nixpkgs/lib/modules.nix":38:17:
while evaluating ‘mergeModules'’ at "/root/nixpkgs/lib/modules.nix":115:36, called from "/root/nixpkgs/lib/modules.nix":112:5:
while evaluating ‘concatMap’ at "/root/nixpkgs/lib/lists.nix":54:18, called from "/root/nixpkgs/lib/modules.nix":151:9:
while evaluating ‘fold’ at "/root/nixpkgs/lib/lists.nix":20:19, called from "/root/nixpkgs/lib/modules.nix":38:38:
while evaluating ‘fold'’ at "/root/nixpkgs/lib/lists.nix":23:15, called from "/root/nixpkgs/lib/lists.nix":27:8:
while evaluating ‘closeModules’ at "/root/nixpkgs/lib/modules.nix":60:27, called from "/root/nixpkgs/lib/modules.nix":34:16:
while evaluating anonymous function at "/root/nixpkgs/lib/modules.nix":62:49, called from "/root/nixpkgs/lib/lists.nix":49:19:
infinite recursion encountered
error: while evaluating the attribute ‘pkgs’ at "/root/nixpkgs/nixos/lib/eval-config.nix":52:3:
while evaluating anonymous function at "/root/nixpkgs/pkgs/top-level/all-packages.nix":8:1, called from "/root/nixpkgs/nixos/lib/eval-config.nix":55:10:
while evaluating ‘applyGlobalOverrides’ at "/root/nixpkgs/pkgs/top-level/all-packages.nix":107:26, called from "/root/nixpkgs/pkgs/top-level/all-packages.nix":101:10:
while evaluating ‘pkgsFun’ at "/root/nixpkgs/pkgs/top-level/all-packages.nix":125:19, called from "/root/nixpkgs/pkgs/top-level/all-packages.nix":120:14:
while evaluating the attribute ‘config.nixpkgs’ at "/root/nixpkgs/nixos/lib/eval-config.nix":26:5:
while evaluating the attribute ‘config’ at "/root/nixpkgs/lib/modules.nix":56:25:
while evaluating ‘yieldConfig’ at "/root/nixpkgs/lib/modules.nix":43:29, called from "/root/nixpkgs/lib/modules.nix":42:16:
while evaluating ‘mapAttrs’ at "/root/nixpkgs/lib/attrsets.nix":134:17, called from "/root/nixpkgs/lib/modules.nix":44:32:
while evaluating ‘mergeModules’ at "/root/nixpkgs/lib/modules.nix":111:26, called from "/root/nixpkgs/lib/modules.nix":38:17:
while evaluating ‘mergeModules'’ at "/root/nixpkgs/lib/modules.nix":115:36, called from "/root/nixpkgs/lib/modules.nix":112:5:
while evaluating ‘concatMap’ at "/root/nixpkgs/lib/lists.nix":54:18, called from "/root/nixpkgs/lib/modules.nix":151:9:
while evaluating ‘fold’ at "/root/nixpkgs/lib/lists.nix":20:19, called from "/root/nixpkgs/lib/modules.nix":38:38:
while evaluating ‘fold'’ at "/root/nixpkgs/lib/lists.nix":23:15, called from "/root/nixpkgs/lib/lists.nix":27:8:
while evaluating ‘closeModules’ at "/root/nixpkgs/lib/modules.nix":60:27, called from "/root/nixpkgs/lib/modules.nix":34:16:
while evaluating anonymous function at "/root/nixpkgs/lib/modules.nix":62:49, called from "/root/nixpkgs/lib/lists.nix":49:19:
infinite recursion encountered
building the system configuration...
error: while evaluating the attribute ‘config.system.build.toplevel’ at "/root/nixpkgs/nixos/lib/eval-config.nix":26:5:
while evaluating anonymous function at "/root/nixpkgs/pkgs/top-level/all-packages.nix":8:1, called from "/root/nixpkgs/nixos/lib/eval-config.nix":55:10:
while evaluating ‘applyGlobalOverrides’ at "/root/nixpkgs/pkgs/top-level/all-packages.nix":107:26, called from "/root/nixpkgs/pkgs/top-level/all-packages.nix":101:10:
while evaluating ‘pkgsFun’ at "/root/nixpkgs/pkgs/top-level/all-packages.nix":125:19, called from "/root/nixpkgs/pkgs/top-level/all-packages.nix":120:14:
while evaluating the attribute ‘config.nixpkgs’ at "/root/nixpkgs/nixos/lib/eval-config.nix":26:5:
while evaluating the attribute ‘config’ at "/root/nixpkgs/lib/modules.nix":56:25:
while evaluating ‘yieldConfig’ at "/root/nixpkgs/lib/modules.nix":43:29, called from "/root/nixpkgs/lib/modules.nix":42:16:
while evaluating ‘mapAttrs’ at "/root/nixpkgs/lib/attrsets.nix":134:17, called from "/root/nixpkgs/lib/modules.nix":44:32:
while evaluating ‘mergeModules’ at "/root/nixpkgs/lib/modules.nix":111:26, called from "/root/nixpkgs/lib/modules.nix":38:17:
while evaluating ‘mergeModules'’ at "/root/nixpkgs/lib/modules.nix":115:36, called from "/root/nixpkgs/lib/modules.nix":112:5:
while evaluating ‘concatMap’ at "/root/nixpkgs/lib/lists.nix":54:18, called from "/root/nixpkgs/lib/modules.nix":151:9:
while evaluating ‘fold’ at "/root/nixpkgs/lib/lists.nix":20:19, called from "/root/nixpkgs/lib/modules.nix":38:38:
while evaluating ‘fold'’ at "/root/nixpkgs/lib/lists.nix":23:15, called from "/root/nixpkgs/lib/lists.nix":27:8:
while evaluating ‘closeModules’ at "/root/nixpkgs/lib/modules.nix":60:27, called from "/root/nixpkgs/lib/modules.nix":34:16:
while evaluating anonymous function at "/root/nixpkgs/lib/modules.nix":62:49, called from "/root/nixpkgs/lib/lists.nix":49:19:
infinite recursion encountered
```

**Comment by domenkozar** [comment]
cc @nbp

**Comment by wizeman** [comment]
@KaiSforza Have you read about the non-backwards compatible changes in #6794 (in the top post)?

I was bit by the same issue but I solved it after I applied the fixes mentioned there (specifically, the fixes mentioned in the section `mkOption, mkIf`).

**Comment by KaiSforza** [comment]
I had not really seen that. Will test and see what happens. It seems like it has to do with the amazon ec2 configurations that are included. I'll be looking around there for a bit now.

**Comment by KaiSforza** [comment]
I have checked that this works in a VM that is not using the amazon linux configurations. Looks like that will need to be updated before this stuff will build.

Moving things, removing the modulePath parts and changing some other stuff gets it to work just fine. Closing this.

**Comment by nbp** [comment]
@KaiSforza , Thanks for reporting this issue.
The amazon issue should be fix by commit 9647a72.
